src/Orleans.Core/Statistics/EnvironmentStatisticsBase.cs

```diff
@@ -0,0 +1,109 @@
+using System;
+using System.Threading.Tasks;
+using System.Threading;
+using Microsoft.Extensions.Logging;
+using System.Diagnostics.Metrics;
+using Orleans.Runtime;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Orleans.Statistics;
+
+#nullable enable
+
+/// <summary>
+/// Base class for environment statistics
+/// </summary>
+[SuppressMessage("CodeQuality", "IDE0052:Remove unread private members", Justification = "Makes memory dump debugging easier")]
+internal abstract class EnvironmentStatisticsBase<T> : IHostEnvironmentStatistics, ILifecycleObserver, IDisposable
+    where T : EnvironmentStatisticsBase<T>
+{
+    private CancellationTokenSource? _cts;
+    private Task? _monitorTask;
+
+    private readonly ObservableCounter<long> _availableMemoryCounter;
+    private readonly ObservableCounter<long> _totalPhysicalMemoryCounter;
+
+    protected const float OneKiloByte = 1024;
+
+    protected ILogger _logger;
+    protected TimeSpan _monitorPeriod = TimeSpan.FromSeconds(5);
+
+    /// <inheritdoc />
+    public float? CpuUsage { get; protected set; }
+    /// <inheritdoc />
+    public long? TotalPhysicalMemory => GC.GetGCMemoryInfo().TotalAvailableMemoryBytes;
+    /// <inheritdoc />
+    public long? AvailableMemory { get; protected set; }
+
+    protected EnvironmentStatisticsBase(ILoggerFactory loggerFactory)
+    {
+        _logger = loggerFactory.CreateLogger<T>();
+
+        _availableMemoryCounter = Instruments.Meter.CreateObservableCounter(InstrumentNames.RUNTIME_MEMORY_AVAILABLE_MEMORY_MB, () => (long)(AvailableMemory ?? 0 / OneKiloByte / OneKiloByte), unit: "MB");
+        _totalPhysicalMemoryCounter = Instruments.Meter.CreateObservableCounter(InstrumentNames.RUNTIME_MEMORY_TOTAL_PHYSICAL_MEMORY_MB, () => (long)(TotalPhysicalMemory ?? 0 / OneKiloByte / OneKiloByte), unit: "MB");
+    }
+
+    protected abstract Task Monitor(CancellationToken cancellationToken);
+
+    public async Task OnStart(CancellationToken cancellationToken)
+    {
+        _logger.LogTrace("Starting {Statistics}", typeof(T).Name);
+
+        _cts = new CancellationTokenSource();
+        cancellationToken.Register(() => _cts.Cancel());
```

Comment:
This should be disposed, eg:
```suggestion
        using var _ = cancellationToken.Register(() => _cts.Cancel());
```